File: server/internal/ipc/fixtures_test.go
package ipc

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// fixturesDir points at the cross-language contract fixtures in
// /protocol/fixtures, shared with the Flutter client's own Dart tests.
func fixturesDir(t *testing.T) string {
	t.Helper()
	dir, err := filepath.Abs(filepath.Join("..", "..", "..", "protocol", "fixtures"))
	if err != nil {
		t.Fatalf("resolve fixtures dir: %v", err)
	}
	if _, err := os.Stat(dir); err != nil {
		t.Fatalf("fixtures dir %q not found: %v", dir, err)
	}
	return dir
}

func loadFixture(t *testing.T, name string) []byte {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(fixturesDir(t), name)) //nolint:gosec // fixture path is built from a fixed test-local directory, not external input
	if err != nil {
		t.Fatalf("read fixture %q: %v", name, err)
	}
	return data
}

func TestFixtures_EnvelopeRoundTrip(t *testing.T) {
	t.Parallel()
	names := []string{
		"request_servers_list.json",
		"response_servers_list_success.json",
		"response_error_not_found.json",
		"response_error_version_mismatch.json",
		"request_runs_start.json",
		"response_runs_start_success.json",
		"event_run_started.json",
		"event_run_step_changed.json",
		"event_run_finished.json",
	}
	for _, name := range names {
		t.Run(name, func(t *testing.T) {
			t.Parallel()
			data := loadFixture(t, name)

			var env Envelope
			if err := json.Unmarshal(data, &env); err != nil {
				t.Fatalf("unmarshal fixture: %v", err)
			}
			if env.Type == "" {
				t.Error("fixture is missing a \"type\" field")
			}

			// Re-marshal and unmarshal again: the envelope shape must be
			// stable under a round trip, which is what both Go and Dart
			// implementations are validated against.
			reEncoded, err := json.Marshal(env)
			if err != nil {
				t.Fatalf("re-marshal envelope: %v", err)
			}
			var roundTripped Envelope
			if err := json.Unmarshal(reEncoded, &roundTripped); err != nil {
				t.Fatalf("unmarshal re-marshaled envelope: %v", err)
			}
			if roundTripped.Type != env.Type || roundTripped.ID != env.ID || roundTripped.Event != env.Event {
				t.Errorf("round trip changed envelope identity: got %+v, want %+v", roundTripped, env)
			}
		})
	}
}

func TestFixtures_RequestServersList_DecodesAsRequest(t *testing.T) {
	t.Parallel()
	var env Envelope
	if err := json.Unmarshal(loadFixture(t, "request_servers_list.json"), &env); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if env.Type != MessageRequest || env.Command != CmdServersList || env.ProtocolVersion != ProtocolVersion {
		t.Errorf("envelope = %+v, want a servers.list request at the current protocol version", env)
	}
}

func TestFixtures_ResponseServersListSuccess_PayloadDecodes(t *testing.T) {
	t.Parallel()
	var env Envelope
	if err := json.Unmarshal(loadFixture(t, "response_servers_list_success.json"), &env); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if env.Success == nil || !*env.Success {
		t.Fatal("expected success = true")
	}
	var resp ListServersResponse
	if err := UnmarshalPayload(env.Payload, &resp); err != nil {
		t.Fatalf("decode payload: %v", err)
	}
	if len(resp.Servers) != 1 || resp.Servers[0].Name != "prod" {
		t.Errorf("Servers = %+v, want one server named %q", resp.Servers, "prod")
	}
	if resp.Servers[0].HasCredential != true {
		t.Error("expected has_credential = true")
	}
}

func TestFixtures_ErrorResponses_HaveStructuredError(t *testing.T) {
	t.Parallel()
	for _, tt := range []struct {
		fixture  string
		wantCode string
	}{
		{"response_error_not_found.json", "NOT_FOUND"},
		{"response_error_version_mismatch.json", "PROTOCOL_VERSION_MISMATCH"},
	} {
		t.Run(tt.fixture, func(t *testing.T) {
			t.Parallel()
			var env Envelope
			if err := json.Unmarshal(loadFixture(t, tt.fixture), &env); err != nil {
				t.Fatalf("unmarshal: %v", err)
			}
			if env.Success == nil || *env.Success {
				t.Fatal("expected success = false")
			}
			if env.Error == nil || env.Error.Code != tt.wantCode {
				t.Errorf("Error = %+v, want code %q", env.Error, tt.wantCode)
			}
		})
	}
}

func TestFixtures_RunEvents_DecodePayloads(t *testing.T) {
	t.Parallel()

	t.Run("run.started", func(t *testing.T) {
		t.Parallel()
		var env Envelope
		if err := json.Unmarshal(loadFixture(t, "event_run_started.json"), &env); err != nil {
			t.Fatalf("unmarshal: %v", err)
		}
		var payload RunStartedEvent
		if err := UnmarshalPayload(env.Payload, &payload); err != nil {
			t.Fatalf("decode payload: %v", err)
		}
		if payload.Run.Status != "RUNNING" {
			t.Errorf("Run.Status = %q, want RUNNING", payload.Run.Status)
		}
	})

	t.Run("run.stepChanged", func(t *testing.T) {
		t.Parallel()
		var env Envelope
		if err := json.Unmarshal(loadFixture(t, "event_run_step_changed.json"), &env); err != nil {
			t.Fatalf("unmarshal: %v", err)
		}
		var payload RunStepChangedEvent
		if err := UnmarshalPayload(env.Payload, &payload); err != nil {
			t.Fatalf("decode payload: %v", err)
		}
		if payload.Step.Type != "ARCHIVE" || payload.Step.Status != "SUCCESS" {
			t.Errorf("Step = %+v, want ARCHIVE/SUCCESS", payload.Step)
		}
	})

	t.Run("run.finished", func(t *testing.T) {
		t.Parallel()
		var env Envelope
		if err := json.Unmarshal(loadFixture(t, "event_run_finished.json"), &env); err != nil {
			t.Fatalf("unmarshal: %v", err)
		}
		var payload RunFinishedEvent
		if err := UnmarshalPayload(env.Payload, &payload); err != nil {
			t.Fatalf("decode payload: %v", err)
		}
		if payload.Run.Status != "SUCCESS" || payload.Run.ArchiveSize == 0 {
			t.Errorf("Run = %+v, want a successful run with a non-zero archive size", payload.Run)
		}
	})
}

File: apps/client/test/core/ipc/fixtures_test.dart
// Validates that Envelope and the DTOs in models.dart correctly parse
// (and round-trip) every fixture in /protocol/fixtures — the wire contract
// shared with the Go service (see server/internal/ipc and
// /protocol/README.md).
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:path/path.dart' as p;
import 'package:vps_backup_manager/core/ipc/envelope.dart';
import 'package:vps_backup_manager/core/ipc/models.dart';

String _readFixture(String name) {
  final path = p.join(
    Directory.current.path,
    '..',
    '..',
    'protocol',
    'fixtures',
    name,
  );
  return File(path).readAsStringSync();
}

void main() {
  group('request fixtures', () {
    test('request_servers_list.json', () {
      final env = Envelope.fromJsonString(
        _readFixture('request_servers_list.json'),
      );
      expect(env.isRequest, isTrue);
      expect(env.id, 'req-1');
      expect(env.protocolVersion, kProtocolVersion);
      expect(env.command, 'servers.list');
      expect(env.payload, isNull);
    });

    test('request_runs_start.json', () {
      final env = Envelope.fromJsonString(
        _readFixture('request_runs_start.json'),
      );
      expect(env.isRequest, isTrue);
      expect(env.command, 'runs.start');
      final payload = env.payload! as Map<String, dynamic>;
      expect(payload['job_id'], 'b1c2d3e4f5061728394a5b6c7d8e9f01');
    });
  });

  group('response fixtures', () {
    test('response_servers_list_success.json parses into ServerDto list', () {
      final env = Envelope.fromJsonString(
        _readFixture('response_servers_list_success.json'),
      );
      expect(env.isResponse, isTrue);
      expect(env.success, isTrue);
      final payload = env.payload! as Map<String, dynamic>;
      final servers = (payload['servers'] as List)
          .map((e) => ServerDto.fromJson(e as Map<String, dynamic>))
          .toList();
      expect(servers, hasLength(1));
      final server = servers.single;
      expect(server.id, '5f9e1c2b7a3d4e6f8091a2b3c4d5e6f7');
      expect(server.name, 'prod');
      expect(server.host, '203.0.113.5');
      expect(server.port, 22);
      expect(server.authType, 'PRIVATE_KEY');
      expect(server.hasCredential, isTrue);
      expect(server.hasTrustedHostKey, isTrue);
    });

    test('response_runs_start_success.json', () {
      final env = Envelope.fromJsonString(
        _readFixture('response_runs_start_success.json'),
      );
      expect(env.success, isTrue);
      final payload = env.payload! as Map<String, dynamic>;
      expect(payload['run_id'], 'c2d3e4f5061728394a5b6c7d8e9f0112');
      expect(payload['status'], 'RUNNING');
    });

    test('response_error_not_found.json', () {
      final env = Envelope.fromJsonString(
        _readFixture('response_error_not_found.json'),
      );
      expect(env.isResponse, isTrue);
      expect(env.success, isFalse);
      expect(env.error, isNotNull);
      expect(env.error!.code, 'NOT_FOUND');
      expect(env.error!.message, 'job "missing-id" not found');
    });

    test('response_error_version_mismatch.json', () {
      final env = Envelope.fromJsonString(
        _readFixture('response_error_version_mismatch.json'),
      );
      expect(env.success, isFalse);
      expect(env.error!.code, 'PROTOCOL_VERSION_MISMATCH');
    });
  });

  group('event fixtures', () {
    test('event_run_started.json parses into RunDto', () {
      final env = Envelope.fromJsonString(
        _readFixture('event_run_started.json'),
      );
      expect(env.isEvent, isTrue);
      expect(env.event, 'run.started');
      final payload = env.payload! as Map<String, dynamic>;
      final run = RunDto.fromJson(payload['run'] as Map<String, dynamic>);
      expect(run.id, 'c2d3e4f5061728394a5b6c7d8e9f0112');
      expect(run.jobId, 'b1c2d3e4f5061728394a5b6c7d8e9f01');
      expect(run.sourcePaths, ['/docker/volumes']);
      expect(run.trigger, 'MANUAL');
      expect(run.status, 'RUNNING');
      expect(run.isTerminal, isFalse);
      expect(run.recoveryOutcome, 'NOT_APPLICABLE');
    });

    test('event_run_step_changed.json parses into StepDto', () {
      final env = Envelope.fromJsonString(
        _readFixture('event_run_step_changed.json'),
      );
      expect(env.event, 'run.stepChanged');
      final payload = env.payload! as Map<String, dynamic>;
      expect(payload['run_id'], 'c2d3e4f5061728394a5b6c7d8e9f0112');
      final step = StepDto.fromJson(payload['step'] as Map<String, dynamic>);
      expect(step.id, 'd3e4f5061728394a5b6c7d8e9f011223');
      expect(step.type, 'ARCHIVE');
      expect(step.status, 'SUCCESS');
      expect(step.truncated, isFalse);
      expect(step.durationMs, 247000);
    });

    test('event_run_finished.json parses into a terminal RunDto', () {
      final env = Envelope.fromJsonString(
        _readFixture('event_run_finished.json'),
      );
      expect(env.event, 'run.finished');
      final payload = env.payload! as Map<String, dynamic>;
      final run = RunDto.fromJson(payload['run'] as Map<String, dynamic>);
      expect(run.status, 'SUCCESS');
      expect(run.isTerminal, isTrue);
      expect(run.archiveName, 'beresta_2026-09-08_03-00-00.tar.gz');
      expect(run.archiveSize, 445123456);
      expect(
        run.checksum,
        '9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08',
      );
    });
  });

  group('Envelope.request round-trip', () {
    test('produces the same shape as request_runs_start.json', () {
      final env = Envelope.request('req-4', 'runs.start', {
        'job_id': 'b1c2d3e4f5061728394a5b6c7d8e9f01',
      });
      final decoded = jsonDecode(env.toJsonString()) as Map<String, dynamic>;
      final expected = jsonDecode(
        _readFixture('request_runs_start.json'),
      ) as Map<String, dynamic>;
      expect(decoded, expected);
    });
  });
}
